src/System.Private.ServiceModel/src/System/ServiceModel/Channels/CommunicationObject.cs

```diff
@@ -37,6 +39,13 @@ protected CommunicationObject(object mutex)
             _mutex = mutex;
             _eventSender = this;
             _state = CommunicationState.Created;
+
+            // All WCF product types fully implement IAsyncCommunicationObject.
+            // But because the interface is internal, custom CommunicationObjects cannot override
+            // any of its methods.  The heuristic to detect this is based on the type's namespace
+            // belonging to System.ServiceModel.
+            string ns = this.GetType().Namespace;
+            _canUseIAsyncCommunicationObject = !String.IsNullOrEmpty(ns) && ns.StartsWith("System.ServiceModel");
```

Comment:
Reviewers: after discussion with @mconnew, we decided on this approach to detect "external" CommunicationObjects.  Unfortunately, IAsyncCommunicationObject is internal and external customers cannot override its methods.  But because our base classes implement it, we can't distinguish simply by testing whether the type implements IAsyncCommunicationObject -- because that would be true even for external types.  And we wanted to avoid a deeper probe into MethodInfo's or we would need to include metadata in NET Native.

I can add deeper Debug-only-on-CoreCLR testing if we want to verify this logic is enough.